### PR TITLE
Make OpenAI API the default API

### PR DESCRIPTION
**Pull request by oobabooga**
The most popular LLM API in the world is the OpenAI API, so I think that it makes sense to emulate it in this project when the `--api` flag is provided. This is already the case for [vLLM](https://github.com/vllm-project/vllm) and [FastChat](https://github.com/lm-sys/FastChat).

The goal is to start from the current `openai` extension and make the following changes before merging:

- [x] Replace the current `http.server` implementation with FastAPI, such that the API docs can be accessed at `127.0.0.1:5001/docs`.
- [x] Use [sse-starlette](https://github.com/sysid/sse-starlette) for streaming.
- [x] Find some sensible and transparent way to handle instruction templates by model.
- [x] The same for truncation lengths.
- [x] Make the overall structure more similar to the [AUTOMATIC1111 api](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/api), with the input and output parameters and types for each endpoint declared in classes.
- [x] Include the additional sampling parameters present in this project as optional parameters in the completion endpoints.

### Status

I have converted everything to fastAPI, and the completion endpoints seemingly work (both with and without streaming).
 
### OpenAI API reference

https://platform.openai.com/docs/api-reference

----

cc @matatonic

### CODE REVIEW

**Comment by matatonic** [comment]
@oobabooga 

I think this is awesome, I don't have the time to contribute right now, but this is great! One additional idea I had was to also allow virtual model aliases to be configured via a UI for the API. It could allow mapping of things like "gpt-3.5-turbo" to a full model config (generation params, model file, loading params, etc.) - and allow switching models on the fly easily, like if an API call came in with "code-davinci-002" you could switch to a code completion model.

**Comment by oobabooga** [comment]
That's an interesting idea @matatonic, I'll see if I can come up with something in this direction -- the only caveat is that the user will have to have downloaded an appropriate model. 

This PR will probably take forever to complete, but I think it's very important. Eventually I'll get there.

File: extensions/openai/fastapi (broken).py
Comment (AlpinDale) [inline extensions/openai/fastapi (broken).py:1]: Would it be better to rename this to `protocols.py` for consistency with other projects?

Comment (oobabooga) [reply extensions/openai/fastapi (broken).py:1]: I think that can be done once the refactor is done, yes

**Comment by teddybear082** [comment]
So this is going to sound bananas as someone who almost exclusively uses the openAI API with textgen but I think your core approach of having a core “textgen” API with addons or extensions for other APIs feels like the right one.  On the other hand yes I do think it would be great for the openai API extension for now to be treated as a “first class citizen” of sorts and get a fresh coat of paint.  The reason I would not change everything over to JUST be the openai API is that we are still in 2023, ChatGPT just launched and made waves months ago so of course as the current market leader all current “cool” stuff is being made to utilize it.  But who is to say that will still be the case in 2024?  Meta is in the process of launching Llama on so many services for free and already has the server infrastructure to transmit text to servers at high speed all around the world.  What if 2024 sees meta launch free API access to Llama2 or Llama2.5 or whatever and they have their own API fornat?  Suddenly we may see services crested mostly for that platform as people will not have to pay to use it.  Same for Microsoft, google, apple.  And it’s not like the OpenAI API is the most user friendly - having to deal with all those dictionaries within dictionaries within arrays.  As they continue to push toward multi-modal they could even change their own API again the way they did from completions to chat completions.  So I just worry about you basing the core of your API software on open AI and then it getting deprecated or outdated.  Maybe I am misunderstanding the intent here though - maybe by “one API” you just mean people wouldn’t have to turn on and off different endpoints they would just be available and you could add on to them and all would have access to the core program features and changes made to the GUI, which would be great.


Anyway I’m just one voice and don’t know very much so take this with a grain of salt.  I do think you’ve accomplished something amazing here which is democratizing AI and making it as easy to use as possible.

**Comment by oobabooga** [comment]
@teddybear082 even if another API becomes more popular, using FastAPI, type hints, SSE instead of websockets, and having an API documentation will still be a benefit. It will be a matter of changing the endpoints and parameters while retaining the overall structure.

It will also be possible for extensions to add endpoints very easily by simply importing `app` from `extensions.openai.script`.

**Comment by LoopControl** [comment]
If I may make an early request -- it would be great to have batching support also so we can have multiple parallel streams. llama.cpp has recently merged it into master so something along those lines would be great.

**Comment by Ph0rk0z** [comment]
My main gripes with OAI api is the mentioned lack of instruction templates and the lack of sampling parameters. Not sure if that's the fault of how the front ends implement API support or the API itself.

**Comment by oobabooga** [comment]
More polishing is needed, but I think that this is starting to make sense. The instruction template can be passed as an optional parameter; if unset, it is detected from `models/config.yaml`. The truncation length is set from the model metadata. The docs can be found at `http://127.0.0.1/docs`.

Any test of the API would be welcome.

**Comment by jsboige** [comment]
Any chance to keep the legacy API for some time? 
The raw deletion of all legacy files over night seems brutal for me. I'll be happy to migrate, but I spent quite some time efficiently supporting websockets and the many parameters allowed in the Oobabooga connector for semantic-kernel, and I'm sure some others have also built a strong dependency on the legacy API, possibly leveraging parameters like generation and characters presets that won't be trivial to migrate. 
Overall, I think it is a good move though, and I'll be happy to update the semantic-fleet connector to move the OpenAI API, but it would be nice if you leave us some time for that with both APIs available, especially since OpenAI won't support many parameters that the legacy offers. 
Funnily enough the semantic-kernel team has taken the inverse step of removing the default OpenAI parameters to make it more flexible to support other LLMs. I guess they haven't really followed the trends outside of the OpenAI/Azure ecosystem, they could have known better and added an OpenAI API to the copilot-chat application instead.

**Comment by teddybear082** [comment]
> More polishing is needed, but I think that this is starting to make sense. The instruction template can be passed as an optional parameter; if unset, it is detected from `models/config.yaml`. The truncation length is set from the model metadata. The docs can be found at `http://127.0.0.1/docs`.
> 
> Any test of the API would be welcome.

I can test; I assume we just pull down this branch and then follow normal install instructions right?

**Comment by LoopControl** [comment]
Agreed with @jsboige -- this is a pretty significant breaking change because of the complete nuking of the existing API which many are probably dependent on.

One popular program is SillyTavern that has explicit support for the text-gen api so some kind of deprecation notice on the existing API for at least a week or 2 before merging this to master would be best so others have time to start switching to the new API.

**Comment by tidymonkey81** [comment]
i setup this pr and am getting this on my server logs (using a previously functioning simple autogen script). not sure but seems FastAPI related?:

`ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 274, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\extensions\openai\script.py", line 107, in openai_chat_completions
    response = OAIcompletions.chat_completions(to_dict(request_data), is_legacy=is_legacy)
                                               ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tidym\git\lab\text-generation-webui\extensions\openai\typing.py", line 141, in to_dict
    return obj.model_dump()
           ^^^^^^^^^^^^^^
AttributeError: 'ChatCompletionRequest' object has no attribute 'model_dump'`

**Comment by teddybear082** [comment]
I checked out the openai branch, installed using start windows.bat.  I got an error about not being able to have two things access the same port, but restating textgenwebui solved that. All basic functionalities appeared to work including  downloading model and running on cpu as well as GPU layers (I chose cuda 12 in the installer step).  I chose api and share and then restarted. I could access the public link and generation worked. However I could not figure out how to access the docs in the addresses set forth above and couldn’t get my openai tests to work (there was also no message in the command window about an API being available at a certain port).  Will try again later on today and read the code more.

**Comment by oobabooga** [comment]
@jsboige @teddybear082 you are right, I have added the original API back with a warning. Once this PR lands into main, I'll update the deprecation date.

> i setup this pr and am getting this on my server logs (using a previously functioning simple autogen script):

I couldn't reproduce this bug. It looks like you are using the chat completions endpoint without streaming; here is a test:

```python
import openai

openai.api_key = "..."
openai.api_base = "http://127.0.0.1:5000/v1"
openai.api_version = "2023-05-15"

chat_completion = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=[{"role": "user", "content": "What is your name?"}], stream=False)
print(chat_completion)
```

**Comment by teddybear082** [comment]
I can cooberate the same error as above. 

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 274, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\script.py", line 107, in openai_chat_completions
    response = OAIcompletions.chat_completions(to_dict(request_data), is_legacy=is_legacy)
                                               ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\typing.py", line 141, in to_dict
    return obj.model_dump()
           ^^^^^^^^^^^^^^
AttributeError: 'ChatCompletionRequest' object has no attribute 'model_dump'   
```
My code that generates this error is: 

```
import openai
openai.api_key = "sk-test"
openai.api_base = "http://127.0.0.1:5000/v1"

completion = openai.ChatCompletion.create(
  model="gpt-3.5-turbo",
  messages=[
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Hello!"}
  ]
)
print("whole response: \n")
print(completion)
print("\nMessage content:\n")
print(completion.choices[0].message)
```

Also it looks like when I use commandflags.txt I just needed api set to create the API but in the user interface, just setting the API flag last time and apply and restart did not generate the openai api.  Going to test the streaming test next.

**Comment by teddybear082** [comment]
Error with streaming test is as follows: 

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 69, in app
    await response(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\sse_starlette\sse.py", line 233, in __call__
    async with anyio.create_task_group() as task_group:
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\anyio\_backends\_asyncio.py", line 597, in __aexit__
    raise exceptions[0]
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\sse_starlette\sse.py", line 236, in wrap
    await func()
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\sse_starlette\sse.py", line 221, in stream_response
    async for data in self.body_iterator:
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\script.py", line 100, in generator
    response = OAIcompletions.stream_chat_completions(to_dict(request_data), is_legacy=is_legacy)
                                                      ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\typing.py", line 141, in to_dict
    return obj.model_dump()
           ^^^^^^^^^^^^^^
AttributeError: 'ChatCompletionRequest' object has no attribute 'model_dump'
```

code that generates the error is: 

```
import openai
import time
# Example of an OpenAI ChatCompletion request with stream=True
# https://platform.openai.com/docs/guides/chat
openai.api_key = "sk-test"
openai.api_base = "http://127.0.0.1:5000/v1"
# record the time before the request is sent
start_time = time.time()

# send a ChatCompletion request to count to 100
response = openai.ChatCompletion.create(
    model='gpt-3.5-turbo',
    messages=[
        {'role': 'user', 'content': 'Count to 100, with a comma between each number and no newlines. E.g., 1, 2, 3, ...'}
    ],
    temperature=0,
    stream=True  # again, we set stream=True
)

# create variables to collect the stream of chunks
collected_chunks = []
collected_messages = []
# iterate through the stream of events
for chunk in response:
    print(chunk)
    chunk_time = time.time() - start_time  # calculate the time delay of the chunk
    collected_chunks.append(chunk)  # save the event response
    chunk_message = chunk['choices'][0]['delta']  # extract the message
    collected_messages.append(chunk_message)  # save the message
    print(f"Message received {chunk_time:.2f} seconds after request: {chunk_message}")  # print the delay and text

# print the time delay and text received
print(f"Full response received {chunk_time:.2f} seconds after request")
full_reply_content = ''.join([m.get('content', '') for m in collected_messages])
print(f"Full conversation received: {full_reply_content}")
```

**Comment by teddybear082** [comment]
In case helpful I get this traceback on my program's side: 

```
Traceback (most recent call last):
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_requestor.py", line 413, in handle_error_response
    error_data = resp["error"]
                 ~~~~^^^^^^^^^
TypeError: string indices must be integers, not 'str'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Willi\Documents\GitHub\koboldcpp\openaichattest.py", line 5, in <module>
    completion = openai.ChatCompletion.create(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_resources\chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_resources\abstract\engine_api_resource.py", line 155, in create
    response, _, api_key = requestor.request(
                           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_requestor.py", line 299, in request
    resp, got_stream = self._interpret_response(result, stream)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_requestor.py", line 710, in _interpret_response
    self._interpret_response_line(
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_requestor.py", line 775, in _interpret_response_line
    raise self.handle_error_response(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\anaconda3\envs\koboldcpp\Lib\site-packages\openai\api_requestor.py", line 415, in handle_error_response
    raise error.APIError(
openai.error.APIError: Invalid response object from API: 'Internal Server Error' (HTTP response code was 500)
```

**Comment by teddybear082** [comment]
I will pull down the new changes but on the version without the last four commits, I wanted to specifically try running your example code above, and got the same error with that code: 

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 274, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\installer_files\env\Lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\script.py", line 107, in openai_chat_completions
    response = OAIcompletions.chat_completions(to_dict(request_data), is_legacy=is_legacy)
                                               ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Willi\Documents\GitHub\text-generation-webui\extensions\openai\typing.py", line 141, in to_dict
    return obj.model_dump()
           ^^^^^^^^^^^^^^
AttributeError: 'ChatCompletionRequest' object has no attribute 'model_dump'
```

will report back once I try with the new commits above.

**Comment by teddybear082** [comment]
ok I pulled down the commits above and still same error. It looks like the code is expecting a method called "model_dump" that would be part of the chat completions class and would turn all the attributes of the class into a dictionary so that then the typing method could turn it into json but that method isn't anywhere to be seen.  sorry wish I could help more.

**Comment by teddybear082** [comment]
sorry for the spam.  changing these lines in typing.py seems to fix it:

```
def to_json(obj):
    #return json.dumps(obj.model_dump(), indent=4)
    return json.dumps(obj__dict__, indent=4)

def to_dict(obj):
    #return obj.model_dump()
    return obj.__dict__
```

**Comment by tidymonkey81** [comment]
> sorry for the spam.  changing these lines in typing.py in extensions / openai seems to fix it, all my tests passed that previously failed with this change.
> 
> ```
> def to_json(obj):
>     #return json.dumps(obj.model_dump(), indent=4)
>     return json.dumps(obj.__dict__, indent=4)
> 
> def to_dict(obj):
>     #return obj.model_dump()
>     return obj.__dict__
> ```
> 
> @tidymonkey81 do you want to test this change on your end and see if it works for you?

This fix did work for me.

**Comment by teddybear082** [comment]
awesome thanks! 

Ooba, I also had some time just now and "stress tested" this by letting it power Mantella, the AI Skyrim mod for NPCs a bit.  Mantella uses the openai api chat completions, both streaming and non-streaming mode for various functions.  Worked well!

**Comment by oobabooga** [comment]
@teddybear082 I have added your fix in the commit above. Thank you so much for the tests @teddybear082 and @tidymonkey81, that really helps.

**Comment by oobabooga** [comment]
I think that the goal of this PR has been achieved -- moving the features of the old API to the OpenAI API. It is passing all my tests.

The docs need some more polishing, but that can be done later. I'll merge the PR so that more people can test the updated API in the dev branch.

**Comment by teddybear082** [comment]
just for completeness here, I did pull down the dev branch today and it also cleared my openai-api tests.

**Comment by Igoorx** [comment]
How do I do to abort the completion? Closing the connection doesn't make the completion stop.

**Comment by oobabooga** [comment]
@Igoorx this was lacking in the OpenAI API. I have added a `/v1/internal/stop-generation` endpoint in this PR: https://github.com/oobabooga/text-generation-webui/pull/4498. Can you check if it works as expected?

**Comment by oobabooga** [comment]
That is indeed a bug. It should be fixed after https://github.com/oobabooga/text-generation-webui/commit/97c21e5667437a706b5aaeb3f6600890003d63c1

If you notice anything else weird, please let me know!

**Comment by teddybear082** [comment]
I don't know if this has any impact on any thing and you may have seen it but on the openai playground, the scripts now recommend a slightly different format to call the openai API in python code.  Perhaps this doesn't impact at all the server side of trying to emulate the openai server, only the requestor, but figured I would pass along just in case.  There's also a new "response format" field that can be passed, maybe something like jsonformers can be used or grammar implemented when someone passes that field, not sure: 

```
# This code is for v1 of the openai package: pypi.org/project/openai
from openai import OpenAI
client = OpenAI()

response = client.chat.completions.create(
  model="gpt-4-1106-preview",
  messages=[
    {
      "role": "user",
      "content": ""
    }
  ],
  temperature=1,
  max_tokens=256,
  top_p=1,
  frequency_penalty=0,
  presence_penalty=0
)
```

https://platform.openai.com/playground?mode=chat&model=gpt-4-1106-preview

https://platform.openai.com/docs/quickstart?context=python

https://platform.openai.com/docs/api-reference/chat/streaming

**Pull request merged**